pkg/controller/master/virtualmachine/vmi_network_controller.go

```diff
@@ -12,107 +13,179 @@ import (
-	vmiClient vmv1.VirtualMachineInstanceClient
-	crCache   ctlappsv1.ControllerRevisionCache
-	crClient  ctlappsv1.ControllerRevisionClient
+	vmCache       vmv1.VirtualMachineCache
+	vmClient      vmv1.VirtualMachineClient
+	crCache       ctlappsv1.ControllerRevisionCache
+	crClient      ctlappsv1.ControllerRevisionClient
+	vmiController vmv1.VirtualMachineInstanceController
 }
 
-// SetDefaultNetworkMacAddress set the default mac address of networks using the initial allocated mac address from the VMI status,
-// since the most guest OS will use the initial allocated mac address of its DHCP config, and on Kubevirt the VM restart it will re-allocate
-// a new mac address which will lead the original network unreachable.
-func (h *VMNetworkController) SetDefaultNetworkMacAddress(id string, vmi *kubevirtv1.VirtualMachineInstance) (*kubevirtv1.VirtualMachineInstance, error) {
-	if id == "" || vmi == nil || vmi.DeletionTimestamp != nil {
+func (h *VMNetworkController) SyncMacAddressAndControllerRevision(id string, vmi *kubevirtv1.VirtualMachineInstance) (*kubevirtv1.VirtualMachineInstance, error) {
+	if id == "" || vmi == nil || vmi.DeletionTimestamp != nil || vmi.Status.Phase != kubevirtv1.Running || len(vmi.Status.Interfaces) == 0 {
 		return vmi, nil
 	}
 
-	if vmi.Status.Phase != kubevirtv1.Running {
-		return vmi, nil
+	// ensure vmi mac is backfilled to vm
+	retryvmi := false
+	if vmi, err := h.syncNetworkDynamicMacAddress(vmi, &retryvmi); err != nil {
+		return vmi, err
+	}
+
+	// ensure vmi mac is backfilled to rc
+	retrycr := false
+	if vmi, err := h.syncControllerRevision(vmi, &retrycr); err != nil {
+		return vmi, err
 	}
 
-	if len(vmi.Status.Interfaces) == 0 {
+	// vmi may end onChange, but vm/rc is not synced yet
+	if retryvmi || retrycr {
+		h.vmiController.EnqueueAfter(vmi.Namespace, vmi.Name, time.Second*1)
 		return vmi, nil
 	}
 
-	err := h.updateVMDefaultNetworkMacAddress(vmi)
+	// last step: if vm still did not sync with rc, update it
+	return h.updateVM(vmi)
+}
+
+// if user does not specify mac address when creating a VM
+// guest OS will use the initial allocated mac address from its DHCP config
+// when VM restarts that mac address may change, and related IP is not persistent
+// syncNetworkDynamicMacAddress saves the dynamic mac address observed from the the VMI status to vm
+func (h *VMNetworkController) syncNetworkDynamicMacAddress(vmi *kubevirtv1.VirtualMachineInstance, retry *bool) (*kubevirtv1.VirtualMachineInstance, error) {
+	vm, err := h.vmCache.Get(vmi.Namespace, vmi.Name)
 	if err != nil {
 		return vmi, err
 	}
 
+	if vm == nil || vm.DeletionTimestamp != nil {
+		return vmi, nil
+	}
+
+	vmCopy := vm.DeepCopy()
+	vmiInterfaces := getVmiInterfaceMacs(vmi)
+	updated := false
+	for i, vmIface := range vmCopy.Spec.Template.Spec.Domain.Devices.Interfaces {
+		if vmIface.Name == "" || vmIface.MacAddress != "" {
+			continue
+		}
+		if macAddress := vmiInterfaces[vmIface.Name]; macAddress != "" {
```

Comment:
do we need the extra check here to verify `macAddress` is not empty?
When we populate the map in `getVMInterfaces` we have already checked macAddress is not empty. It should not be possible to find an empty `macAddress` by the time the map lookup is performed.

Reply:
That's for below case:

1.1 User creates a VM with 2 interfaces, and start the VM
1.2 User adds an additional interface to VM, but do not restart the VM right now
1.3 In this loop, with VM's interface to check VMI's interface map, it may get an empty one.